build.py

```diff
@@ -0,0 +1,418 @@
+                f"{DIR_PATH}/templates",
+            ]
+        )
+
+    def assemble_convert_command(self):
+        return " ".join(
+            [
+                f"{self.qemu_path}",
+                f"convert",
+                f"-O",
+                f"raw",
+                f"./images/{self.template}",
+                f"{self.image_path}",
+            ]
+        )
+
+    def assemble_os_command(self):
+        return " ".join(
+            [
+                f"{self.openstack_path}",
+                f"image",
+                f"create",
+                f"--file",
+                f"{self.image_path}",
+                f"{self.image_name}",
+            ]
+        )
+
+    def assemble_packer_envs(self):
+        env = os.environ.copy()
+        env["PACKER_PLUGIN_PATH"] = f"{DIR_PATH}/packer_plugins"
+        env[
+            "PKR_VAR_groups"
+        ] = f"[{','.join('\"' + x + '\"' for x in self.provisioning)}]"
+        env["PKR_VAR_headless"] = "true"
+        if self.ansible_args:
+            env["PKR_VAR_ansible_extra_args"] = self.ansible_args
+        return env
+
+    def assemble_name(self):
+        """
+        Uses a naming scheme described in\n
+        https://github.com/usegalaxy-eu/vgcn/issues/78
+        """
+        name = ["vgcn"]
+        if "generic" in self.provisioning:
+            provisioning = self.provisioning
+            provisioning.remove("generic")
+            name += [f"+{'+'.join(provisioning)}"]
+        else:
+            name += [f"!generic+{'+'.join(self.provisioning)}"]
+        name += [
+            "-".join(self.template.split("-", 2)[:2]),
+            self.assemble_timestamp(),
+            subprocess.check_output(["git", "rev-parse", "--abbrev-ref", "HEAD"])
+            .decode("ascii")
+            .strip(),
+            subprocess.check_output(["git", "rev-parse", "--short", "HEAD"])
+            .decode("ascii")
+            .strip(),
+        ]
+        if self.comment:
+            name += [self.comment]
+        return "~".join(name)
+
+    def assemble_scp_command(self):
+        return " ".join(
+            [
+                f"scp",
+                f"-i",
+                f"{self.pvt_key}",
+                f"{self.image_path}",
+                f"{SSH_USER}@{SSH_HOST}:{STATIC_DIR}/{os.path.basename(self.image_path)}",
+            ]
+        )
+
+    def assemble_ssh_command(self):
+        return " ".join(
+            [
+                f"ssh",
+                f"-i",
+                f"{self.pvt_key}",
+                f"{SSH_USER}@{SSH_HOST}",
+                f"chmod",
+                f"ugo+r",
+                f"{STATIC_DIR}{os.path.basename(self.image_path)}",
+            ]
+        )
+
+    def assemble_timestamp(self):
+        today = datetime.date.today()
+        seconds_since_midnight = time.time() - time.mktime(today.timetuple())
+        return today.strftime("%Y%m%d") + "~" + str(int(seconds_since_midnight))
```

Comment:
What we discussed was to use the **commit date and time** ([check issue again](https://github.com/usegalaxy-eu/vgcn/issues/78#issue-1952530534)), the point was to find out which image is "newer" with just a glance.

If you want we can add the build time somewhere else (e.g. UNIX timestamp before the comment).

_Bullshit_ solution for **commit date and time**.

```suggestion
        commit_time = subprocess.check_output(["git", "show", "--no-patch", "--format=%ct"]).decode("ascii").strip()
        commit_time = int(commit_time)
        commit_time = datetime.datetime.fromtimestamp(commit_time)
        seconds_since_midnight = int((commit_time - datetime.datetime(commit_time.date().year, commit_time.date().month, commit_time.date().day)).total_seconds())
        return commit_time.date().strftime("%Y%m%d") + "~" + str(int(seconds_since_midnight))
```

Please do not add this as it is, make it less verbose.

Reply:
yes, you are right, it's not like specified in the issue and I will change that.
we should somehow make sure that if someone builds an image locally there needs to be a commit before the build happens.

Reply:
sorry what do you mean by less verbose?

Reply:
I mean that the code I am proposing is not very readable. I assume it can be refactored into something easier to read.

Reply:
```suggestion
        commit_time = subprocess.check_output(
            ["git", "show", "--no-patch", "--format=%ct"]
        ).decode("ascii").strip()
        commit_time = int(commit_time)
        commit_time = datetime.datetime.fromtimestamp(commit_time)

        commit_date = commit_time.date()
        commit_date_midnight = datetime.datetime(
            commit_date.year,
            commit_time.month,
            commit_time.day
        )

        seconds_since_midnight = int(
            (commit_time - commit_date_midnight).total_seconds()
        )

        return (
            f"{commit_time.date().strftime('%Y%m%d')}"
            f"~{seconds_since_midnight}"
        )
```

This one is hard to get readable **and** short. I'll commit this, feel free to improve.